### PR TITLE
feat: typescript

### PR DESCRIPTION
**Pull request by viceice**
uses ts-node to execute

Closes #3402

relates #1261

### CODE REVIEW

**Comment by ViceIce** [comment]
will add typescript linting tomorrow

https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#readme

File: test/datasource/npm/index.spec.js [hunk 1]
```diff
@@ -365,10 +365,12 @@ describe('api/npm', () => {
...
   it('should use global cache', async () => {
-    const dummyValue = 'abc123';
-    await global.renovateCache.set('datasource-npm', 'foobar', dummyValue, 10);
+    const dep = {
+      name: 'abc123',
+    };
+    await global.renovateCache.set('datasource-npm', 'foobar', dep, 10);
     const res = await npm.getPkgReleases({ lookupName: 'foobar' });
-    expect(res).toEqual(dummyValue);
+    expect(res).toEqual(dep);
```
Comment (rarkins) [inline test/datasource/npm/index.spec.js]: Is this change necessary as part of this PR?

File: package.json [hunk 2]
```diff
@@ -203,7 +210,9 @@
     ],
     "snapshotSerializers": [
       "./test/newline-snapshot-serializer.js"
-    ]
+    ],
+    "testPathIgnorePatterns": ["/node_modules/", "/.cache/", "/dist/"],
```
Comment (rarkins) [inline package.json]: Is this necessary as part of this PR, or an unrelated optimization?

**Comment by rarkins** [comment]
TODO:
- [ ] Ensure that semantic-release step has build done prior

File: test/datasource/npm/index.spec.js (hunk 1 above)
Comment (ViceIce) [reply test/datasource/npm/index.spec.js]: I don't now, is from @JamieMagee. I can remove that.

File: package.json (hunk 2 above)
Comment (ViceIce) [reply package.json]: It's an optimization, i can make a seperat pr for it, i had set the baseDir to that, so jest has found some tests in the checked out repositories.

**Comment by JamieMagee** [comment]
@ViceIce thanks for picking this up and driving it 😁. I'm currently on vacation, and don't have access to my dev environment

File: test/datasource/npm/index.spec.js (hunk 1 above)
Comment (ViceIce) [reply test/datasource/npm/index.spec.js]: removed

File: package.json (hunk 2 above)
Comment (ViceIce) [reply package.json]: removed

File: test/datasource/npm/index.spec.js (hunk 1 above)
Comment (ViceIce) [reply test/datasource/npm/index.spec.js]: After reverted test, i found another bug:

```
TypeError: Cannot create property 'tags' on string 'abc123'
      at Object.getPkgReleases (lib/datasource/npm/releases.js:14)
```

The cache object should be an object 😄

Comment (rarkins) [reply test/datasource/npm/index.spec.js]: I'll commit it to master then and merge in here

Comment (ViceIce) [reply test/datasource/npm/index.spec.js]: ok, I've made a seperate commit 😄

Comment (ViceIce) [reply test/datasource/npm/index.spec.js]: #3599

File: Dockerfile [hunk 3]
```diff
@@ -143,8 +143,8 @@ ENV PATH="/home/ubuntu/.yarn/bin:/home/ubuntu/.config/yarn/global/node_modules/.
 COPY package.json .
 COPY yarn.lock .
 RUN yarn install --production && yarn cache clean
-COPY lib lib
+COPY dist dist
```
Comment (rarkins) [inline Dockerfile]: I think this may cause a problem with our automated Docker builds. I need to work out if Docker can actually build the project first before building the Dockerfile. I suspect it can't, which means the Docker images would break.

**Review by rarkins — CHANGES_REQUESTED** [review]
Need to work out the Docker build problem

File: Dockerfile (hunk 3 above)
Comment (ViceIce) [reply Dockerfile]: OK, can we use 2 stage docker build?

Comment (ViceIce) [reply Dockerfile]: then we can in first stage build renovate with node:lts-alpine and then copy dist from stage=0

Comment (rarkins) [reply Dockerfile]: I was thinking like this:
image
From: https://docs.docker.com/docker-hub/builds/advanced/

But your idea might be simpler. Let's build on the same image/platform we use though.

Comment (ViceIce) [reply Dockerfile]: if we use 2 stages we do not have the dev deps installed in the final layer :-)

so i would prefer 2 stages, but we can do it in a single stage too. docker builds supports multistage builds :-)

Comment (ViceIce) [reply Dockerfile]: I will update the dockerfile, then you can see it :-)

Comment (rarkins) [reply Dockerfile]: ok thanks. I *think* I am nearly done, just in case you haven't started yet

Comment (ViceIce) [reply Dockerfile]: OK, i'm done 😄

Comment (ViceIce) [reply Dockerfile]:
```patch
diff --git a/Dockerfile b/Dockerfile
index a9800f62..02f35410 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,3 +1,16 @@
+FROM node:lts-alpine@sha256:313c5c88acc0ec12a9abca9a83719065dfd54e94ffc56464b7ce24998dd2838d
+
+WORKDIR /usr/src/app/
+
+COPY package.json .
+COPY yarn.lock .
+RUN yarn install
+
+COPY lib lib
+
+RUN yarn build
+
+
 FROM amd64/ubuntu:18.04@sha256:f2557f94cac1cc4509d0483cb6e302da841ecd6f82eb2e91dc7ba6cfd0c580ab

 LABEL maintainer="Rhys Arkins <rhys@arkins.net>"
@@ -143,7 +156,7 @@ ENV PATH="/home/ubuntu/.yarn/bin:/home/ubuntu/.config/yarn/global/node_modules/.
 COPY package.json .
 COPY yarn.lock .
 RUN yarn install --production && yarn cache clean
-COPY dist dist
+COPY --from=0 /usr/src/app/dist dist
 COPY bin bin

 ENTRYPOINT ["node", "/usr/src/app/dist/renovate.js"]
```

Comment (ViceIce) [reply Dockerfile]: why we need this:
https://github.com/renovatebot/renovate/blob/ba543a029512a26c17f6437f488c73e1d0a389dd/Dockerfile#L107

Comment (rarkins) [reply Dockerfile]: I forget, but it looks like the type of line not to remove unless you do know why it'd needed :D

Comment (rarkins) [reply Dockerfile]: Potentially chowning it to the ubuntu user would also solve whatever problem that caused it

**Comment by rarkins** [comment]
To be extra cautious, I will treat this as a breaking change i.e. a major version bump to v17.0.0

I've changed the target branch of this PR to be `v17` and will make sure it builds on Docker Hub first before merging to master
